docs/docs/adding-react-components.md

```diff
@@ -0,0 +1,124 @@
+---
+title: Adding React Components
+---
+
+This guide covers how to add React components to your Gatsby site.
+
+## React components
+
+React components are prebuilt elements or groups of elements that can be used to prevent repeating yourself in your code.
+
+Components also offer the ability to be customised using "props" (properties). Props can be of any Javascript type, such as boolean, string, object, array or almost anything you can think of.
+
+For example, you could use a component for Buttons on your site. This is because they would likely be used multiple times across pages and would need different labels each time.
+
+## Importing React components
+
+In Gatsby, when using React components, you import and use them just like you would in normal React:
+
+```jsx
+import React from "react"
+import { Link } from "gatsby"
+
+export default () => (
+  <div>
+    <Link to="/contact/">Contact</Link>
+  </div>
+)
+```
+
+## Things to watch out for
+
+Since Gatsby uses Server Side Rendering (SSR) to generate your site's pages, the JSX code you write is usually compiled before the browser loads the page. Because of this, certain features are not available at compile time and can cause a build to error.
+
+### Use of browser globals
+
+Some components or code references on "browser globals" such as `window`, `document` or `localStorage`. This object is not available at compile time so can result in a Webpack error when building:
+
+```
+WebpackError: ReferenceError: window is not defined
+```
+
+#### Fixing your code
+
+You can use conditionals to check that the browser global is defined in your code:
+
+```js
+if (typeof window !== "undefined") {
+  // code that references the browser global
+  window.alert("Woohoo!")
+}
+```
+
+You can also drop the code that uses the browser global inside [`componentDidMount`](https://reactjs.org/docs/react-component.html#componentdidmount):
+
+```jsx
+import React, { useEffect } from "react"
+
+export default () => {
+  useEffect(() => {
+    // code that references the browser global
+    window.alert("Woohoo!")
+  })
+
+  return (
+    <div>
+      <p>Component</p>
+    </div>
+  )
+}
+```
+
+...or you can use [a `useEffect` hook](https://reactjs.org/docs/hooks-reference.html#useeffect):
+
+```jsx
+import React, { Component } from "react"
+
+class MyComponment() extends Component {
+  componentDidMount() {
+    // code that references the browser global
+    window.alert("Woohoo!")
+  }
+
+  render() {
+    return (
+      <div>
+        <p>Component</p>
+      </div>
+    )
+  }
+}
+```
+
+#### Fixing third-party modules
```

Comment:
This section is great, and looks it comes from the [Debugging HTML Builds page](https://www.gatsbyjs.org/docs/debugging-html-builds/), rather than copying that content here, it would be nice to add a note on this doc to about this warning and link to that content instead. We could remove this section that's already included in that doc, and instead cross reference, which will help keep the docs up to date as these pages get updated in the future.